ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/extensions causes Matrix crash loop on Node 24 (native TS loader skips .js→.ts remapping)

## Summary

The official Docker image `ghcr.io/openclaw/openclaw:2026.4.5` sets `OPENCLAW_BUNDLED_PLUGINS_DIR=/app/extensions`, pointing to the TypeScript **source** directory. On Node.js v24 (bundled in this image), which has native TypeScript support enabled (`process.features.typescript = true`), this causes a crash loop when the Matrix plugin loads: Node 24 picks up `.ts` files natively but does not remap `.js → .ts` for static imports inside those files.

## Steps to Reproduce

1. Run the official image with Matrix enabled (default config):
```bash
docker run ghcr.io/openclaw/openclaw:2026.4.5
```
2. The container enters a crash loop immediately.

## Actual Behavior

Crash loop with:
```
Cannot find module '/app/extensions/matrix/src/account-selection.js'
imported from /app/extensions/matrix/src/matrix/credentials-read.ts
```

**Root cause chain:**
- `OPENCLAW_BUNDLED_PLUGINS_DIR=/app/extensions` → raw TypeScript source
- Node 24 has native TS support active — loads `credentials-read.ts` natively when `credentials-read.js` is requested
- Node 24's native TS loader does **not** remap `.js → .ts` for static imports inside the loaded `.ts` file
- TypeScript convention uses `.js` extensions in import statements (e.g. `import ... from './account-selection.js'`) — Node 24 does not resolve these to their `.ts` counterparts
- `account-selection.js` not found → crash

## Expected Behavior

Bundled plugins should load from `/app/dist/extensions/` (compiled JS), which already exists in the image and does not trigger this issue.

## Environment

| Component | Value |
|-----------|-------|
| Image | `ghcr.io/openclaw/openclaw:2026.4.5` |
| Node.js | v24.14.0 (bundled in image) |
| `process.features.typescript` | `true` |
| Plugin affected | Matrix |
| `OPENCLAW_BUNDLED_PLUGINS_DIR` (current) | `/app/extensions` (TS source) |
| `OPENCLAW_BUNDLED_PLUGINS_DIR` (needed) | `/app/dist/extensions` (compiled JS) |

## Proposed Fix

In the official image Dockerfile, change:
```dockerfile
ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/extensions
```
to:
```dockerfile
ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/dist/extensions
```

`/app/dist/extensions/` is already present in the image with compiled JavaScript plugins.

## Workaround

```dockerfile
FROM ghcr.io/openclaw/openclaw:2026.4.5
ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/dist/extensions
```

Or at runtime:
```bash
docker run -e OPENCLAW_BUNDLED_PLUGINS_DIR=/app/dist/extensions ghcr.io/openclaw/openclaw:2026.4.5
```

## Additional Notes

This does **not** reproduce on Node 22 — JITI handles full resolution there. The bug is triggered by the Node 24 upgrade bundled in the 2026.4.5 image.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -56,6 +56,7 @@ Docs: https://docs.openclaw.ai
 - Telegram/doctor: keep top-level access-control fallback in place during multi-account normalization while still promoting legacy default auth into `accounts.default`, so existing named bots keep inherited allowlists without dropping the legacy default bot. (#62263) Thanks @obviyus.
 - Agents/subagents: honor `sessions_spawn(lightContext: true)` for spawned subagent runs by preserving lightweight bootstrap context through the gateway and embedded runner instead of silently falling back to full workspace bootstrap injection. (#62264) Thanks @theSamPadilla.
 - Slack/media: keep attachment downloads on the SSRF-guarded dispatcher path so Slack media fetching works on Node 22 without dropping pinned transport enforcement. (#62239) Thanks @openperf.
+- Docker/plugins: stop forcing bundled plugin discovery to `/app/extensions` in runtime images so packaged installs use compiled `dist/extensions` artifacts again and Node 24 containers do not boot through source-only plugin entry paths. Fixes #62044.
 
 ## 2026.4.5
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -56,7 +56,7 @@ Docs: https://docs.openclaw.ai
 - Telegram/doctor: keep top-level access-control fallback in place during multi-account normalization while still promoting legacy default auth into `accounts.default`, so existing named bots keep inherited allowlists without dropping the legacy default bot. (#62263) Thanks @obviyus.
 - Agents/subagents: honor `sessions_spawn(lightContext: true)` for spawned subagent runs by preserving lightweight bootstrap context through the gateway and embedded runner instead of silently falling back to full workspace bootstrap injection. (#62264) Thanks @theSamPadilla.
 - Slack/media: keep attachment downloads on the SSRF-guarded dispatcher path so Slack media fetching works on Node 22 without dropping pinned transport enforcement. (#62239) Thanks @openperf.
-- Docker/plugins: stop forcing bundled plugin discovery to `/app/extensions` in runtime images so packaged installs use compiled `dist/extensions` artifacts again and Node 24 containers do not boot through source-only plugin entry paths. Fixes #62044.
+- Docker/plugins: stop forcing bundled plugin discovery to `/app/extensions` in runtime images so packaged installs use compiled `dist/extensions` artifacts again and Node 24 containers do not boot through source-only plugin entry paths. Fixes #62044. (#62316) Thanks @gumadeiras.
 
 ## 2026.4.5
 

--- a/Dockerfile
+++ b/Dockerfile
@@ -160,10 +160,6 @@ COPY --from=runtime-assets --chown=node:node /app/skills ./skills
 COPY --from=runtime-assets --chown=node:node /app/docs ./docs
 COPY --from=runtime-assets --chown=node:node /app/qa ./qa
 
-# In npm-installed Docker images, prefer the copied source extension tree for
-# bundled discovery so package metadata that points at source entries stays valid.
-ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/${OPENCLAW_BUNDLED_PLUGIN_DIR}
-
 # Keep pnpm available in the runtime image for container-local workflows.
 # Use a shared Corepack home so the non-root `node` user does not need a
 # first-run network fetch when invoking pnpm.

--- a/src/dockerfile.test.ts
+++ b/src/dockerfile.test.ts
@@ -50,12 +50,10 @@ describe("Dockerfile", () => {
     );
   });
 
-  it("pins bundled plugin discovery to copied source extensions in runtime images", async () => {
+  it("does not override bundled plugin discovery in runtime images", async () => {
     const dockerfile = await readFile(dockerfilePath, "utf8");
     expect(dockerfile).toContain(`ARG OPENCLAW_BUNDLED_PLUGIN_DIR=${BUNDLED_PLUGIN_ROOT_DIR}`);
-    expect(dockerfile).toContain(
-      "ENV OPENCLAW_BUNDLED_PLUGINS_DIR=/app/${OPENCLAW_BUNDLED_PLUGIN_DIR}",
-    );
+    expect(dockerfile).not.toContain("ENV OPENCLAW_BUNDLED_PLUGINS_DIR=");
   });
 
   it("normalizes plugin and agent paths permissions in image layers", async () => {

--- a/src/dockerfile.test.ts
+++ b/src/dockerfile.test.ts
@@ -7,6 +7,10 @@ import { BUNDLED_PLUGIN_ROOT_DIR } from "../test/helpers/bundled-plugin-paths.js
 const repoRoot = resolve(fileURLToPath(new URL(".", import.meta.url)), "..");
 const dockerfilePath = join(repoRoot, "Dockerfile");
 
+function collapseDockerContinuations(dockerfile: string): string {
+  return dockerfile.replace(/\\\r?\n[ \t]*/g, " ");
+}
+
 describe("Dockerfile", () => {
   it("uses shared multi-arch base image refs for all root Node stages", async () => {
     const dockerfile = await readFile(dockerfilePath, "utf8");
@@ -51,9 +55,9 @@ describe("Dockerfile", () => {
   });
 
   it("does not override bundled plugin discovery in runtime images", async () => {
-    const dockerfile = await readFile(dockerfilePath, "utf8");
+    const dockerfile = collapseDockerContinuations(await readFile(dockerfilePath, "utf8"));
     expect(dockerfile).toContain(`ARG OPENCLAW_BUNDLED_PLUGIN_DIR=${BUNDLED_PLUGIN_ROOT_DIR}`);
-    expect(dockerfile).not.toContain("ENV OPENCLAW_BUNDLED_PLUGINS_DIR=");
+    expect(dockerfile).not.toMatch(/^\s*ENV\b[^\n]*\bOPENCLAW_BUNDLED_PLUGINS_DIR\b/m);
   });
 
   it("normalizes plugin and agent paths permissions in image layers", async () => {

